### PR TITLE
Hey guys! I would like to contribute this program

### PR DESCRIPTION
**Pull request by HCMY**
This project is so interesting!!!

Added missForest, a widely used imputation method. I have been tested this update, including using setup.py! Thanks a lot!

I am doing some research on data completion, I have found some more advanced methods, I look forward to joining you

I am the author of [ycimpute](https://github.com/OpenIDEA-YunanUniversity/ycimpute) and I would like to join you! 
I will follow the coding style of the project

### CODE REVIEW

**Comment by HCMY** [comment]
Due to my new computer doesn't have git currently, so I commit via upload files and I missed some files leads to automated tests failed 2 times. Finally, it passed hahahah.

File: fancyimpute/missforest.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            x_new_imp = np.asarray(x_new_imp)
+
+            differ_continuous, differ_categorical = self._lose_func(x_new_imp, x_old_imp)
+            if differ_continuous >= differ_continuous_old and differ_categorical >= differ_categorical_old:
+                self.stop = True
+        return init_fill
+
+    def _lose_func(self, imp_new, imp_old):
+        """
+        Evaluation Method, mathematical concept are available at 'https://www.stu-zhouyc.com/iterForest/metrics'
```
Comment (sergeyf) [inline fancyimpute/missforest.py]: This website doesn't seem to exist?

File: fancyimpute/missforest.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        try:
+            continuous_div = continuous_imp_new - continuous_imp_old
+            continuous_div = continuous_div.dot(continuous_div)
+            continuous_sum = continuous_imp_new.dot(continuous_imp_new)
+
+            categorical_count = np.sum(categorical_imp_new == categorical_imp_old)
+            categorical_var_len = len(categorical_imp_new)
+
+        except:
```
Comment (sergeyf) [inline fancyimpute/missforest.py]: What kind of errors do you expect here? Can you be more explicit in catching them?

File: test/test_missforest.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+def test_missforest_all_continous():
+    solver = MissForest(n_estimators=500)
+    XY_completed = solver.fit_transform(XY_incomplete) #dont need to normalization
+    _, missing_mae = reconstruction_error(
+        XY,
+        XY_completed,
+        missing_mask,
+        name="MissForest")
+    assert missing_mae < 0.3, "Error too high!"
+'''
```
Comment (sergeyf) [inline test/test_missforest.py]: Why are these commented out?

File: test/test_missforest.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+def test_missforest_all_continous():
+    solver = MissForest(n_estimators=500)
+    XY_completed = solver.fit_transform(XY_incomplete) #dont need to normalization
+    _, missing_mae = reconstruction_error(
+        XY,
+        XY_completed,
+        missing_mask,
+        name="MissForest")
+    assert missing_mae < 0.3, "Error too high!"
```
Comment (sergeyf) [inline test/test_missforest.py]: Every other test has this as `missing_mae < 0.1` which suggests this is too high.

File: fancyimpute/missforest.py (hunk 1 above)
Comment (HCMY) [reply fancyimpute/missforest.py]: Well, thank you for your reply! It looks like scikit-learn already has MICE and missForest implementations. I should not implement these methods anymore.

File: fancyimpute/missforest.py (hunk 2 above)
Comment (HCMY) [reply fancyimpute/missforest.py]: Yes, its moved, I forget this, my bad

File: test/test_missforest.py (hunk 4 above)
Comment (HCMY) [reply test/test_missforest.py]: Ok, thanks

**Comment by HCMY** [comment]
I have been implementd a new method, but I writed in Pytorch On GPU, Coud I contribution on this program?

**Comment by sergeyf** [comment]
@HCMY If you have run this new method on a good set of metrics and it performs well compared to our existing algorithms, then sure.

**Comment by HCMY** [comment]
> @HCMY If you have run this new method on a good set of metrics and it performs well compared to our existing algorithms, then sure.

Thanks! I will try

**Comment by HCMY** [comment]
> @HCMY If you have run this new method on a good set of metrics and it performs well compared to our existing algorithms, then sure.

Do I have to test on the data set produced by "low_rank_data"? Because of some characteristics of the algorithm: such as which need for a part of the complete data set and the data size requirements, I worry that under this data set, the algorithm may not perform better than the library function, But I tested it on the real data set, this method performs well

**Comment by HCMY** [comment]
> > @HCMY If you have run this new method on a good set of metrics and it performs well compared to our existing algorithms, then sure.
> 
> Do I have to test on the data set produced by "low_rank_data"? Because of some characteristics of the algorithm: such as which need for a part of the complete data set for training and the data size requirements, I worry that under this data set, the algorithm may not perform better than the library function, But I tested it on the real data set, this method performs well

**Comment by sergeyf** [comment]
Can you also run our datasets on your real data?

**Comment by HCMY** [comment]
> Can you also run our datasets on your real data?

Yup! I just ran Matrix-Complete and SoftImpute on your dataset, your methods were performance better. Here is the relative paper coms from PAKDD  [Gondara, Lovedeep and Wang, Ke. Mida: Multiple imputation using denoising autoencoders, 2017](https://arxiv.org/abs/1705.02737)

**Comment by sergeyf** [comment]
Ah if our methods are better, it's probably not worth it to add a new one!

**Comment by HCMY** [comment]
> Ah if our methods are better, it's probably not worth it to add a new one!

Yes, I understand this, but when I using your data set to test this new method, there is only 22 complete samples  for training, and last 472 samples need to be complete.  When I tested this on UCI two data set(4W+ sample, named Spam and Wave), this method performs better. Can I continue to make some contributions?

**Comment by sergeyf** [comment]
If you think it will pass the 0.1 mae tests, then go ahead.

**Comment by HCMY** [comment]
> If you think it will pass the 0.1 mae tests, then go ahead.

Thanks, I will try to improve this!

**Comment by sergeyf** [comment]
Closing. Feel free to reopen as needed.

**Pull request closed without merging**